network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/exclusion/ui/AppExclusionListAdapter.kt

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.networkprotection.impl.exclusion.ui
+
+import android.content.Context
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.RecyclerView
+import com.duckduckgo.app.global.extensions.safeGetApplicationIcon
+import com.duckduckgo.mobile.android.ui.view.gone
+import com.duckduckgo.mobile.android.ui.view.show
+import com.duckduckgo.networkprotection.impl.R
+import com.duckduckgo.networkprotection.impl.databinding.ItemExclusionListAppBinding
+import com.duckduckgo.networkprotection.impl.databinding.ItemExclusionListFilterBinding
+import com.duckduckgo.networkprotection.impl.databinding.ItemExclusionListHeaderBinding
+import com.duckduckgo.networkprotection.impl.exclusion.ui.AppsProtectionType.AppType
+import com.duckduckgo.networkprotection.impl.exclusion.ui.AppsProtectionType.FilterType
+import com.duckduckgo.networkprotection.impl.exclusion.ui.AppsProtectionType.HeaderType
+
+class AppExclusionListAdapter(
+    val listener: ExclusionListListener,
+) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+    companion object {
+        private const val HEADER_TYPE = 0
+        private const val FILTER_TYPE = 1
+        private const val APP_TYPE = 2
+        private const val HEADER_ITEMS = 3
+    }
+
+    private var isListEnabled: Boolean = false
+    private val exclusionListItems = mutableListOf<AppsProtectionType>()
+
+    fun update(
+        viewState: ViewState,
+        isListStateEnabled: Boolean = true,
+    ) {
+        isListEnabled = isListStateEnabled
+        val oldData = exclusionListItems
+        val newList = viewState.apps
+        val diffResult = DiffCallback(oldData, newList).run { DiffUtil.calculateDiff(this) }
+        exclusionListItems.clear().also { exclusionListItems.addAll(newList) }
+        diffResult.dispatchUpdatesTo(this)
+    }
+
+    override fun getItemCount(): Int {
+        return exclusionListItems.size
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return when (exclusionListItems[position]) {
+            is HeaderType -> HEADER_TYPE
+            is FilterType -> FILTER_TYPE
+            else -> APP_TYPE
+        }
+    }
+
+    override fun onCreateViewHolder(
```

Comment:
Dunno about your code style but in case the whole method is wrapped with `when` or `with` it could be a lot flatten and IMO readable:

```kotiln
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): RecyclerView.ViewHolder = when (viewType) {
        HEADER_TYPE -> HeaderViewHolder.create(parent)
        FILTER_TYPE -> FilterViewHolder.create(parent)
        else -> AppViewHolder.create(parent)
    }
```